### PR TITLE
Add contributing Guide

### PR DESCRIPTION
**Pull request by BoisterousCoder**
I am adding a contributing guide

### CODE REVIEW

File: contributing.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Contributing to {{Project-Name}}
+
+## Where to get help
+The main way to get help is on our [discord server](https://discord.gg/uh69TdKfBD).  Though, this guide should help you get started. It may be slightly lengthy, but it designed for those who are new so please don't let length intimidate you.
+
+## Code of Conduct
+Please be nice to each other when interacting with others and follow our [code of conduct](<link to code of conduct>).
+
+## How to contribute
+In order to contribute you can either add a new issue or you can create a pull request and link it to an existing issue. If the code adds a feature that is not already present in an issue, you can create a new issue for the feature and add the pull request to it.
```
Comment (bgins) [inline contributing.md:10]: > In order to contribute you can either add a new issue or you can create a pull request and ~~link it to~~ reference it in an existing issue.

Possible slight rewording

> If the code adds a feature that is not already present in an issue, you can create a new issue for the feature and add the pull request to it.

We may want to encourage folks to create an issue before they start writing any code. The concern is that if someone does a bunch of work without discussing it first, and their pull request is declined, they may feel bad about the experience. Instead, if there is a discussion in an issue, they can get a sense for whether their work will be accepted.

Probably less of a concern for fixing bugs, but it's still a good idea to discuss it a bit first.

File: contributing.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+### Contributing by adding an issue
+If you have found a bug and would like to report it or if you have a feature that you feel we should add, then it would be appreciated if you contributed by adding an issue. Before you do so please search the other issues to ensure that you are not attempting to add an issue that has already been created. In order to submit a new issue just hit the issue button and a choice between two templates should appear and just follow along with the template you chose. If you don't know how to fill in all parts of the template just go ahead and skip those parts. You can edit the issue later.
```
Comment (bgins) [inline contributing.md:13]: > ~~If you have found a bug and would like to report it or if you have a feature that you feel we should add, then it would be appreciated if you contributed by adding an issue.~~ If you have found a bug or would like to request a feature, we'd love it if you opened an issue! ❤️ 

Possible rewording, a bit shorter and with a positive vibe. I'd also consider a paragraph break after this line.

> Before you do, ~~so~~ please search the ~~other~~ existing issues to ~~ensure that you are not attempting to add an issue that has already been created.~~ avoid creating a duplicate issue.

Possible rewording.

> ~~In order to~~ To submit a new issue, ~~just~~ hit

Maybe reword this a bit. @zeeshanlakhani is probably appreciating this by now, since he knows my editing advice usually involves removing a lot of words. 😄 

> ~~and just follow~~ Follow along with the template you chose.

Consider making this a new sentence. Shorter sentences are easier to read.

> If you don't know how to fill in all parts of the template, ~~just~~ go ahead and skip those parts. You can edit the issue later.

"just" can be omitted here.

File: contributing.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+### Contributing through code
+In order to contribute through code follow the steps below. Note that you don't need to be the best programmer to contribute. Our discord is open for questions
+
+ 1. Pick a feature you would like to add or a bug you would like to fix.
```
Comment (bgins) [inline contributing.md:18]: If a bullet point has a single sentence, it doesn't need punctuation at the end. As one technical writing instructor put it to me in the past, the bullet point is your punctuation.

File: contributing.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+### Contributing by adding an issue
+If you have found a bug and would like to report it or if you have a feature that you feel we should add, then it would be appreciated if you contributed by adding an issue. Before you do so please search the other issues to ensure that you are not attempting to add an issue that has already been created. In order to submit a new issue just hit the issue button and a choice between two templates should appear and just follow along with the template you chose. If you don't know how to fill in all parts of the template just go ahead and skip those parts. You can edit the issue later.
+
+### Contributing through code
+In order to contribute through code follow the steps below. Note that you don't need to be the best programmer to contribute. Our discord is open for questions
+
+ 1. Pick a feature you would like to add or a bug you would like to fix. 
+	 - If you wish to contribute but what you want to fix/add is not already listed in the issues go ahead and add the issue
```
Comment (bgins) [inline contributing.md:19]: > If you wish to contribute but what you want to fix/add is not already ~~listed in the issues go ahead and add the issue~~ covered in an existing issue, please open a new issue 

Suggested rewording.

Also, we may want another bullet point here encouraging folks to discuss the bug or feature before they start coding.

File: contributing.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ 2. Create a fork of the repository
+	 - A fork creates your own version of the code on your Github, so you can work on it separately from everyone else.
```
Comment (bgins) [inline contributing.md:21]: > A fork creates ~~your own version of the code on your Github~~ a copy of the code in your GitHub account

Possible rewording.

File: contributing.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	 - You can learn more about forking [here](https://docs.github.com/en/get-started/quickstart/fork-a-repo)
+ 3. Ensure that you have commit signing enabled
+	 - This ensures that the code you submit is made by you and not someone else say that they are you
```
Comment (bgins) [inline contributing.md:24]: > This ensures that the code you submit ~~is made~~ was committed by you and not someone ~~say that they are you~~ who claims to be you

Suggested rewording.

File: contributing.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	 - You can learn more about how to setup commit signing [here](https://www.freecodecamp.org/news/what-is-commit-signing-in-git/ "https://www.freecodecamp.org/news/what-is-commit-signing-in-git/")
+	 - If you have already made some commits that you wish to put in a pull request without signing them, then you can follow [this guide](https://dev.to/jmarhee/signing-existing-commits-with-gpg-5b58) on how to fix that.
```
Comment (bgins) [inline contributing.md:26]: Yes! Great to have these links in here 💯

File: contributing.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ 5. Clone the repository to your your computer.
+	 - This puts a copy of your fork of the code on your computer so you can edit it
```
Comment (bgins) [inline contributing.md:28]: > This puts a copy of your fork ~~of the code~~ on your computer so you can edit it

"of the code" is implied and can be omitted.

File: contributing.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+	 - You can learn more about cloning repositories [here](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository)
+ 6. Build the project
+	 - For a detailed look on how to build {{project-name}} look at our [README file](<link to readme>)
+ 7. Now is the time to start writing the code you wanted
+	 - Just fire up your favorite code editor and make the changes that you wanted to make to the repository
```
Comment (bgins) [inline contributing.md:33]: > Just fire up your favorite code editor and make the changes that you wanted to make to the repository

Love the energy here, but it could be intimidating to folks who are newer to programming. I think the "Just" could be interpreted as "it's so easy!". Being a bit nitpicky here and maybe reading too much into it.

One alternative way you could open this sentence is with "Open your favorite code editor...".

File: contributing.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+	 - Make sure to test your code with `{{test-code-cmd}}`
+ 8. Please ensure that the code that you made follows our rust coding guide lines
```
Comment (bgins) [inline contributing.md:35]: > Please ensure that the code ~~that you made~~ you wrote follows our rust coding guide lines

Slight rewording.

File: contributing.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+	 - You can find our list of guidelines [here](<link to coding guidelines>)
+	 - This is a kindness to the programmers that come after you. The easier your code is to read, the easier it will be for the next person to make modifications.
```
Comment (bgins) [inline contributing.md:37]: > This is a ~~kindness~~ courtesy

Could replace kindness with courtesy

File: contributing.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	 - Your code doesn't need to follow all the guide lines all the time but please do try your best to follow them as close as possible
```
Comment (bgins) [inline contributing.md:38]: We might want to leave this out. We can work with folks when they submit a PR to get things up to the guidelines, but we want things to be in order before we merge. @zeeshanlakhani thoughts?

File: contributing.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+ 9. Commit and Push your code
+	 - This sends your code to your repository branch
```
Comment (bgins) [inline contributing.md:40]: > This sends your ~~code~~ changes to your repository branch

Maybe "changes" here instead? The idea is that you are just sending the new bits of code that you worked on.

File: contributing.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+	 - You can learn more about committing code [here](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/making-changes-in-a-branch/committing-and-reviewing-changes-to-your-project), and pushing it [here](https://docs.github.com/en/get-started/using-git/pushing-commits-to-a-remote-repository).
```
Comment (bgins) [inline contributing.md:41]: > , and

The comma can be removed here.

File: contributing.md [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+ 10. The final step is to create pull request to our main branch 🥳🎉
```
Comment (bgins) [inline contributing.md:42]: Yay! Emoji! 🕺

File: contributing.md [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+	 - A pull request is how you merge the code you just worked so hard on with the code everyone else has access to
+	 - Once you have submitted your pull request we will review your code and check to make sure it works. We may leave some feedback and suggest edits. You can make the changes we suggest by simply committing more code to your fork.
```
Comment (bgins) [inline contributing.md:44]:
```suggestion
	 - Once you have submitted your pull request, we will review your code and check to make sure it works. We may leave some feedback and suggest edits. You can make the changes we suggest by committing more code to your fork.
```
Totally forgot about GH suggestions! Should have been using them all along.

> make sure it works

This is a bit ambiguous. Might be worth calling out that the code should implement the feature or fix the bug in the associated issue.

File: contributing.md [hunk 17]
```diff
@@ -0,0 +1,59 @@
+# Contributing to {{Project-Name}}
+
+We welcome everyone to contribute what they can. Whether you are brand new, just want to contribute a little bit, or want to contribute a lot there is probably something you can help out with.
```
Comment (zeeshanlakhani) [inline contributing.md:3]: We should note that we'll provide labels on issues for `good_first_issue` items (https://build.prestashop-project.org/news/a-definition-of-the-good-first-issue-label), making it easy to get started.

File: contributing.md [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+## Where to get help
+The main way to get help is on our [discord server](https://discord.gg/uh69TdKfBD).  Though, this guide should help you get started. It may be slightly lengthy, but it's designed for those who are new so please don't let length intimidate you.
```
Comment (zeeshanlakhani) [inline contributing.md:6]: double-checking, is the link here an infinite invite to the general lobby/channel?

File: contributing.md [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+## Code of Conduct
+Please be nice when interacting with others and follow our [code of conduct](<link to code of conduct>).
```
Comment (zeeshanlakhani) [inline contributing.md:9]: the tl;dr we've been using is `Be kind, inclusive, and considerate.`

So, `Please be kind, inclusive, and considerate when interacting`...

File: contributing.md [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+## How to contribute
+If the code adds a feature that is not already present in an issue, you can create a new issue for the feature and add the pull request to it. If the code adds a feature that is not already present in an issue, you can create a new issue for the feature and add the pull request to it.
```
Comment (zeeshanlakhani) [inline contributing.md:12]: Bring up the difference between GH issue and GH discussion here? (w/ links), right @bgins?

File: contributing.md [hunk 21]
```diff
@@ -0,0 +1,59 @@
+# Contributing to {{Project-Name}}
```
Comment (zeeshanlakhani) [inline contributing.md:1]: minor: since we have a solid amount of links here, which should use the non-inline version of linking (`[foo][link]`), https://commonmark.org/help/tutorial/07-links.html

File: contributing.md [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+### Contributing by adding an issue
+If you have found a bug and would like to report it or if you have a feature that you feel we should add, then we'd love it if you opened an issue! ❤️ Before you do, please search the other issues to avoid creating a duplicate issue. 
+To submit a new issue just hit the issue button and a choice between two templates should appear. Then, follow along with the template you chose. If you don't know how to fill in all parts of the template go ahead and skip those parts. You can edit the issue later.
+
+### Contributing through code
+In order to contribute through code follow the steps below. Note that you don't need to be the best programmer to contribute. Our discord is open for questions
+
+ 1. **Pick a feature** you would like to add or a bug you would like to fix
+	 - If you wish to contribute but what you want to fix/add is not already covered in an existing issue, please open a new issue
```
Comment (zeeshanlakhani) [inline contributing.md:22]: minor: And @bgins knows I'll pick these out, but we should probably use `.` (periods) consistently through these, as they are more like sentences.

File: contributing.md (hunk 21 above)
Comment (zeeshanlakhani) [inline contributing.md:1]: for templating purposes in `cargo-generate`, it'll be `{{project-name}}`.

File: contributing.md (hunk 19 above)
Comment (zeeshanlakhani) [inline contributing.md:9]: minor: `Code of Conduct` vs `code of conduct`. 

Also, since we'll add the CoC to generated repos, the link here can be always relative: `./code_of_conduct.md` or something like this.

File: contributing.md (hunk 20 above)
Comment (zeeshanlakhani) [reply contributing.md:12]: Maybe it's:

```
### Contributing By Adding an Issue
...
### Contributing by Adding a Topic for Discussion?
```

File: contributing.md (hunk 21 above)
Comment (zeeshanlakhani) [inline contributing.md:1]: minor: throughout the doc, should we move capitalize words, instead of lowercase (other than prepositions, etc)?

File: contributing.md (hunk 17 above)
Comment (zeeshanlakhani) [reply contributing.md:3]: We can maybe even embed the label. I'll make sure we have it at the org level.

File: contributing.md [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+## How to contribute
+If the code adds a feature that is not already present in an issue, you can create a new issue for the feature and add the pull request to it. If the code adds a feature that is not already present in an issue, you can create a new issue for the feature and add the pull request to it.
+
+### Contributing by adding an issue
+If you have found a bug and would like to report it or if you have a feature that you feel we should add, then we'd love it if you opened an issue! ❤️ Before you do, please search the other issues to avoid creating a duplicate issue. 
+To submit a new issue just hit the issue button and a choice between two templates should appear. Then, follow along with the template you chose. If you don't know how to fill in all parts of the template go ahead and skip those parts. You can edit the issue later.
+
+### Contributing through code
+In order to contribute through code follow the steps below. Note that you don't need to be the best programmer to contribute. Our discord is open for questions
+
```
Comment (zeeshanlakhani) [inline contributing.md:20]: We should add a thing about kinda following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) here, but can be in another PR too.

File: contributing.md [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+ 1. **Pick a feature** you would like to add or a bug you would like to fix
+	 - If you wish to contribute but what you want to fix/add is not already covered in an existing issue, please open a new issue
+ 2. **Discuss** the issue with the rest of the community
```
Comment (zeeshanlakhani) [inline contributing.md:23]: Yeah, the question here is a discussion on the issue or via a GH discussion: https://docs.github.com/en/discussions. I give @bgins the power to make the best partition for our cases :).

File: contributing.md [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+	 - Before you write any code, it is recommended that you discuss your intention to write the code on the issue you are attempting to edit
+	 -  This helps to stop you from wasting your time duplicating the work of others that maybe working on the same issue; at the same time
+	 - This step also allows you to get helpful pointers on the community on some problems they may have encountered on similar issues
+ 3. **Fork** the repository
+	 - A fork creates a copy of the code on your Github, so you can work on it separately from everyone else.
+	 - You can learn more about forking [here](https://docs.github.com/en/get-started/quickstart/fork-a-repo)
+ 4. Ensure that you have **commit signing** enabled
+	 - This ensures that the code you submit was committed by you and not someone else who claims to be you
+	 - You can learn more about how to setup commit signing [here](https://www.freecodecamp.org/news/what-is-commit-signing-in-git/ "https://www.freecodecamp.org/news/what-is-commit-signing-in-git/")
```
Comment (zeeshanlakhani) [inline contributing.md:32]: Great having this and the links 🥳 @BoisterousCoder.

File: contributing.md [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+	 - If you have already made some commits that you wish to put in a pull request without signing them, then you can follow [this guide](https://dev.to/jmarhee/signing-existing-commits-with-gpg-5b58) on how to fix that.
+ 5. **Clone** the repository to your your computer.
```
Comment (zeeshanlakhani) [inline contributing.md:34]: minor: `your your`

```suggestion
5. **Clone** the repository to your local computer.
```

File: contributing.md [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+	 - This puts a copy of your fork on your computer so you can edit it
+	 - You can learn more about cloning repositories [here](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository)
+ 6. **Build** the project
+	 - For a detailed look on how to build {{project-name}} look at our [README file](<link to readme>)
+ 7. **Start writing** the code you wanted
```
Comment (zeeshanlakhani) [inline contributing.md:39]:
```suggestion
7. **Start writing** your code.
```

File: contributing.md [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+	 - Open up your favorite code editor and make the changes that you wanted to make to the repository
+	 - Make sure to test your code with `{{test-code-cmd}}`
```
Comment (zeeshanlakhani) [inline contributing.md:41]: This can't really be templated per, but we can link to readme here.

File: contributing.md [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+ 8. **Write tests** for your code
```
Comment (zeeshanlakhani) [inline contributing.md:42]: We should talk about when to have a `unit` vs `integration` test: https://doc.rust-lang.org/book/ch11-03-test-organization.html as well and how that relates to Rust.

File: contributing.md [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+	 - If you are adding a new feature you should write tests that ensure that if someone make changes to the code it cannot break your new feature without breaking the test
+	 - If your code adds a new feature you should also write at least one documentation test. The documentation test's purpose is to demonstrate and document how to use the API feature
+	 - If your code fixes a bug you should write tests that ensure that if someone makes code changes in the future the bug does not re-emerge without breaking tests
+	 - {% if wasm %}Tests should be put in the test folder and in a test file relevant to the project you are working in {% end if %}
```
Comment (zeeshanlakhani) [inline contributing.md:46]: on me: since wasm or not is chosen at the project level, we won't if wasm or not. Though if you look at the example(s) in https://github.com/fission-codes/rust-template/pull/3/files#diff-2f1b77c3844317bffd439878f2e6fd1ded5c07fac384a1c85c04cd9e80390856, these links may not be necessary?

File: contributing.md [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+	 - {% if !wasm %}Tests should be put inside the test folder inside the source folder and in a test file relevant to the project you are working in {% end if %}
+	 - For more information on how to write tests look at [this link](https://doc.rust-lang.org/book/ch11-01-writing-tests.html)
+ 9. Ensure that the code that you made follows our rust **coding guide lines**
+	 - You can find our list of guidelines [here](<link to coding guidelines>)
```
Comment (zeeshanlakhani) [inline contributing.md:50]: minor: relative link here too. I'll create an issue for us to start this doc, just a light one to start. 

Also our list of `Rust` guidelines specifically.

File: contributing.md [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+	 - This is a courtesy to the programmers that come after you. The easier your code is to read, the easier it will be for the next person to make modifications.
```
Comment (zeeshanlakhani) [inline contributing.md:51]: no `-`? Does it continue from before?

File: contributing.md [hunk 33]
```diff
@@ -0,0 +1,59 @@
...
+	 - If you find it difficult to follow the guidelines or if the guidelines or unclear please reach out to us through our discord linked above, or you can just continue and leave a comment in the pull request stage.
+ 10. **Commit and Push** your code
+	 - This sends your changes to your repository branch
+	 - You can learn more about committing code [here](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/making-changes-in-a-branch/committing-and-reviewing-changes-to-your-project) and pushing it [here](https://docs.github.com/en/get-started/using-git/pushing-commits-to-a-remote-repository).
```
Comment (zeeshanlakhani) [inline contributing.md:55]: ah, maybe mention the conventional commits here.

**Review by zeeshanlakhani — CHANGES_REQUESTED** [review]
@BoisterousCoder great work thus far! I've added a ton of small comments through this.

File: contributing.md (hunk 21 above)
Comment (zeeshanlakhani) [inline contributing.md:1]: minor: `Rust` for `rust` where applicable.

File: contributing.md [hunk 34]
```diff
@@ -0,0 +1,59 @@
...
+	 - Make sure to test your code with `{{test-code-cmd}}`
+ 8. **Write tests** for your code
+	 - If you are adding a new feature you should write tests that ensure that if someone make changes to the code it cannot break your new feature without breaking the test
+	 - If your code adds a new feature you should also write at least one documentation test. The documentation test's purpose is to demonstrate and document how to use the API feature
+	 - If your code fixes a bug you should write tests that ensure that if someone makes code changes in the future the bug does not re-emerge without breaking tests
+	 - {% if wasm %}Tests should be put in the test folder and in a test file relevant to the project you are working in {% end if %}
+	 - {% if !wasm %}Tests should be put inside the test folder inside the source folder and in a test file relevant to the project you are working in {% end if %}
+	 - For more information on how to write tests look at [this link](https://doc.rust-lang.org/book/ch11-01-writing-tests.html)
+ 9. Ensure that the code that you made follows our rust **coding guide lines**
```
Comment (zeeshanlakhani) [inline contributing.md:49]: Ensure is probably strong here, but we should use the tone of recommend. Not sure how to best to say right here.

Comment (zeeshanlakhani) [reply contributing.md:49]: Also `guide lines` -> `guidelines`.

File: contributing.md (hunk 18 above)
Comment (BoisterousCoder) [reply contributing.md:6]: Yes, it is

File: contributing.md (hunk 20 above)
Comment (bgins) [reply contributing.md:12]: Ah yeah, that's right! We were going to add a bit about discussions vs issues. 💯 

I think that roughly speaking issues are for when you have a clearly defined bug report or feature request. Anything else probably falls under discussions. Examples might include:

- Community announcements and asking the community for feedback
- Discussing a new release
- Asking questions, Q&A that isn't for sure a bug report

There are some other ones listed in this post: https://resources.github.com/devops/process/planning/discussions/

One goal here is keeping issues focused on work that can be tied to PR. That way issues don't linger and stay open for extended periods of time.

File: contributing.md (hunk 22 above)
Comment (bgins) [reply contributing.md:22]: My understanding is that a bullet point that has a single sentence should not have a period at the end. However, if it has more than one sentence it should. As my technical writing instructor put it, the bullet point is your punctuation if you only have one sentence.

File: contributing.md (hunk 24 above)
Comment (bgins) [reply contributing.md:23]: Yep, discussions vs issues here: https://github.com/fission-codes/rust-template/pull/4/files#r1015784880

Generally, I'd say the use cases for discussions are less about contributing through code. So this section seems fine as is -- we still want to discuss in issues.

File: contributing.md (hunk 32 above)
Comment (BoisterousCoder) [reply contributing.md:51]: I am sorry, I am not sure what you mean here. If you mean  `The easier your code is to read, the easier it will be for the next person to make modifications`. Then it definitely seems like a single Item to me

Comment (zeeshanlakhani) [reply contributing.md:51]: @BoisterousCoder sorry for the confusion. Upon re-read, I meant something more like this:

```markdown
9. Ensure that the code that you made follows our rust **coding guide lines**. 

   - You can find our list of guidelines [here](<link to coding guidelines>). This is 
     a courtesy to the programmers that come after you. The easier your code is to read, 
     the easier it will be for the next person to make modifications.
```

Mainly, it seems this can read as one note for #9 because it's all referencing the link?

File: contributing.md [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+## Code of Conduct
+Please be kind, inclusive, and considerate when interacting when interacting with others and follow our [code of conduct](./code_of_conduct.md).
+
+## How to Contribute
+If the code adds a feature that is not already present in an issue, you can create a new issue for the feature and add the pull request to it. If the code adds a feature that is not already present in an issue, you can create a new issue for the feature and add the pull request to it.
+
+### Contributing by Adding a Topic for Discussion
+#### Issues
+If you have found a bug and would like to report it or if you have a feature that you feel we should add, then we'd love it if you opened an issue! ❤️ Before you do, please search the other issues to avoid creating a duplicate issue. 
+To submit a new issue just hit the issue button and a choice between two templates should appear. Then, follow along with the template you chose. If you don't know how to fill in all parts of the template go ahead and skip those parts. You can edit the issue later.
```
Comment (bgins) [inline contributing.md:17]: Could we add a newline above this to separate out the paragraphs?

File: contributing.md [hunk 36]
```diff
@@ -0,0 +1,70 @@
...
+
+#### Discussion
+If you have some new discussion you want to start but it isn't a bug or feature add, then you can start a discussion. Some examples of what kinds of things that are good discussion topics can include, but not limited to the following
```
Comment (bgins) [inline contributing.md:20]:
```suggestion
If you have a new discussion you want to start but it isn't a bug or feature add, then you can start a [GitHub discussion](https://docs.github.com/en/discussions). Some examples of what kinds of things that are good discussion topics can include, but are not limited to the following
```

GitHub discussions are fairly new, so it might be good to link out to the docs for them.

File: contributing.md [hunk 37]
```diff
@@ -0,0 +1,70 @@
...
+-   Asking questions, Q&A that isn't for sure a bug report
+
+### Contributing through Code
+In order to contribute through code follow the steps below. Note that you don't need to be the best programmer to contribute. Our discord is open for questions
+
+ 1. **Pick a feature** you would like to add or a bug you would like to fix
+	 - If you wish to contribute but what you want to fix/add is not already covered in an existing issue, please open a new issue
+ 2. **Discuss** the issue with the rest of the community
+	 - Before you write any code, it is recommended that you discuss your intention to write the code on the issue you are attempting to edit
+	 -  This helps to stop you from wasting your time duplicating the work of others that maybe working on the same issue; at the same time
```
Comment (bgins) [inline contributing.md:32]:
```suggestion
	 - This helps to stop you from wasting your time duplicating the work of others that maybe working on the same issue; at the same time
```

Extra space.

File: contributing.md [hunk 38]
```diff
@@ -0,0 +1,70 @@
...
+	 - This step also allows you to get helpful pointers on the community on some problems they may have encountered on similar issues
+ 3. **Fork** the repository
+	 - A fork creates a copy of the code on your Github, so you can work on it separately from everyone else.
```
Comment (bgins) [inline contributing.md:35]:
```suggestion
	 - A fork creates a copy of the code on your Github, so you can work on it separately from everyone else
```

No period needed at the end of this one.

File: contributing.md [hunk 39]
```diff
@@ -0,0 +1,70 @@
...
+	 - This puts a copy of your fork on your computer so you can edit it
+	 - You can learn more about cloning repositories [here](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository)
+ 6. **Build** the project
+	 - For a detailed look on how to build {{project-name}} look at our [README file](<./README>)
+ 7. **Start writing** your code
+	 - Open up your favorite code editor and make the changes that you wanted to make to the repository
+	 - Make sure to test your code with the test command found in our [README file](<./README>)
+ 8. **Write tests** for your code
+	 - If you are adding a new feature you should write tests that ensure that if someone make changes to the code it cannot break your new feature without breaking the test
+	 - If your code adds a new feature you should also write at least one documentation test. The documentation test's purpose is to demonstrate and document how to use the API feature
```
Comment (bgins) [inline contributing.md:51]:
```suggestion
	 - If your code adds a new feature you should also write at least one documentation test. The documentation test's purpose is to demonstrate and document how to use the API feature.
```

Because there are two sentences in the bullet point, both should have a period at the end.

File: contributing.md [hunk 40]
```diff
@@ -0,0 +1,70 @@
...
+	 - If your code fixes a bug you should write tests that ensure that if someone makes code changes in the future the bug does not re-emerge without breaking test
```
Comment (bgins) [inline contributing.md:52]:
```suggestion
	- If your code fixes a bug you should write tests that ensure that if someone makes code changes in the future the bug does not re-emerge without breaking test
```
Looks like this one was out of alignment by one space.

File: contributing.md [hunk 41]
```diff
@@ -0,0 +1,70 @@
...
+	- Please create integration tests, if the addition is large enough to warrant them, and unit tests
+		 - Unit tests are tests that ensure the functionality of a single function or small section of code
+		 - Integration tests test large large sections of code
+		 - read more about the differences [here](https://www.geeksforgeeks.org/difference-between-unit-testing-and-integration-testing/)
```
Comment (bgins) [inline contributing.md:56]:
```suggestion
		 - Read more about the differences [here](https://www.geeksforgeeks.org/difference-between-unit-testing-and-integration-testing/)
```

Capitalize for consistency with the other bullet points.

File: contributing.md [hunk 42]
```diff
@@ -0,0 +1,70 @@
...
+	 - Tests should be put inside the test folder inside the source folder and in a test file relevant to the project you are working in
+	 - For more information on how to write tests look at [this link](https://doc.rust-lang.org/book/ch11-01-writing-tests.html)
+ 9. Ensure that the code that you made follows our Rust **coding guidelines**
+	 - You can find our list of guidelines [here](./coding-guide)
+	 - This is a courtesy to the programmers that come after you. The easier your code is to read, the easier it will be for the next person to make modifications.
+	 - If you find it difficult to follow the guidelines or if the guidelines or unclear please reach out to us through our discord linked above, or you can just continue and leave a comment in the pull request stage.
```
Comment (bgins) [inline contributing.md:62]:
```suggestion
	 - If you find it difficult to follow the guidelines or if the guidelines or unclear please reach out to us through our discord linked above, or you can just continue and leave a comment at the pull request stage.
```

**Review by bgins — APPROVED** [review]
Looks great! Thanks for all of the work on this! 🙌 🎉 

I left a few copy edit suggestions. Feel free to merge when you've had a look at them.

File: contributing.md [hunk 43]
```diff
@@ -0,0 +1,70 @@
...
+	- If your code fixes a bug you should write tests that ensure that if someone makes code changes in the future the bug does not re-emerge without breaking test
+	- Please create integration tests, if the addition is large enough to warrant them, and unit tests
+		- Unit tests are tests that ensure the functionality of a single function or small section of code
+		- Integration tests test large large sections of code
+		- Read more about the differences [here](https://www.geeksforgeeks.org/difference-between-unit-testing-and-integration-testing/)
+	- Tests should be put inside the test folder inside the source folder and in a test file relevant to the project you are working in
+	- For more information on how to write tests look at [this link](https://doc.rust-lang.org/book/ch11-01-writing-tests.html)
+ 9. Ensure that the code that you made follows our Rust **coding guidelines**
+	- You can find our list of Rust guidelines [here](./coding-guide) This is a courtesy to the programmers that come after you. The easier your code is to read, the easier it will be for the next person to make modifications.
```
Comment (zeeshanlakhani) [inline contributing.md:61]: while we our working on ours, can we link to https://rust-lang.github.io/api-guidelines/about.html?

**Review by zeeshanlakhani — APPROVED** [review]
one last link comment @BoisterousCoder and :shipit:. I'll do some follow-up after this to get it through in through template and the like for cleanup.
